### PR TITLE
Expose Druid functions in `INFORMATION_SCHEMA.ROUTINES` table.

### PR DESCRIPTION
**Pull request by abhishekrb19**
This PR adds a new table `INFORMATION_SCHEMA.ROUTINES`. Resolves https://github.com/apache/druid/issues/14346.

### Description

 The implementation currently exposes information about all functions in `INFORMATION_SCHEMA.ROUTINES`. The table contains the following columns:
   * ROUTINE_CATALOG: Name of the database. Always set to druid.
   * ROUTINE_SCHEMA: Name of the schema. Always set to INFORMATION_SCHEMA.
   * ROUTINE_NAME: Name of the function or operator. E.g., APPROX_COUNT_DISTINCT_DS_THETA
   * ROUTINE_TYPE: Type of routine. Always set to FUNCTION.
   * IS_AGGREGATOR: Is the routine an aggregator or not. Returns YES for aggregator functions; NO for scalar functions.
   * SIGNATURES: Possible signatures for the routine as a string.

Since Calcite is the source of truth for all the registered functions, including ones loaded through custom extensions, the implementation uses the Calcite's operator table to extract this information at runtime.

A preview of the table from the Druid console:

CleanShot 2023-06-09 at 13 09 02@2x

#### Release note

Expose Druid functions via `INFORMATION_SCHEMA.ROUTINES` table.

<hr>

##### Key changed/added classes in this PR
 * `InformationSchema.java`
 * `DruidOperatorTable.java`
 * `InformationSchemaTest.java`
 * `DruidOperatorTableTest.java`

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by FrankChen021** [comment]
I'm wondering why is not the `IS_AGGREGATOR` declared as `BOOLEAN`? 

IMO for boolean type, it would be more simplier and intuitive for user to filter out aggregator functions like 

```sql
select * from ROUTINES where IS_AGGREGATOR
```

**Comment by FrankChen021** [comment]
OK, let's keep it consistent with existing designs.

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/InformationSchema.java [hunk 1]
```diff
@@ -469,6 +487,66 @@ public Object[] apply(final RelDataTypeField field)
...
+    }
+
+    @Override
+    public Enumerable<Object[]> scan(DataContext root)
+    {
+      final List<Object[]> rows = new ArrayList<>();
+      List<SqlOperator> operatorList = operatorTable.getOperatorList();
+
+      for (SqlOperator sqlOperator : operatorList) {
+        if (sqlOperator.getSyntax() != SqlSyntax.FUNCTION &&
```
Comment (zachjsh) [inline sql/src/main/java/org/apache/druid/sql/calcite/schema/InformationSchema.java:528]: Curious, how did you come up with this list of sqlSyntax types to filter against? I see that the org.apache.druid.sql.calcite.planner.DruidOperatorTable#normalizeSyntax seems to treat anything that is not `PREFIX`, `BINARY`, or `POSTFIX` as `FUNCTION`?

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/InformationSchema.java [hunk 2]
```diff
@@ -131,6 +139,14 @@ public RelDataType build()
       .add("COLLATION_NAME", SqlTypeName.VARCHAR, true)
       .add("JDBC_TYPE", SqlTypeName.BIGINT)
       .build();
+  private static final RelDataType ROUTINES_SIGNATURE = new RowTypeBuilder()
+      .add("ROUTINE_CATALOG", SqlTypeName.VARCHAR)
+      .add("ROUTINE_SCHEMA", SqlTypeName.VARCHAR)
+      .add("ROUTINE_NAME", SqlTypeName.VARCHAR)
+      .add("ROUTINE_TYPE", SqlTypeName.VARCHAR)
+      .add("IS_AGGREGATOR", SqlTypeName.VARCHAR)
+      .add("SIGNATURES", SqlTypeName.VARCHAR)
```
Comment (zachjsh) [inline sql/src/main/java/org/apache/druid/sql/calcite/schema/InformationSchema.java:148]: seems that this column can be nullable based on logic below.

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java [hunk 3]
```diff
@@ -429,6 +431,22 @@ public void testAggregatorsOnInformationSchemaColumns()
     );
   }
 
+  @Test
+  public void testAggregatorOnInformationSchemaRoutines()
```
Comment (zachjsh) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java:435]: Should we add a test for count of non-aggregators?

File: sql/src/test/java/org/apache/druid/sql/calcite/schema/InformationSchemaTest.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+    for (Object[] row : rows) {
+      Assert.assertEquals(rowType.getFieldCount(), row.length);
+      Assert.assertEquals("druid", row[0]);
+      Assert.assertEquals("INFORMATION_SCHEMA", row[1]);
+      Assert.assertNotNull(row[2]);
+      Assert.assertNotNull(row[3]);
+      String isAggregator = row[4].toString();
+      Assert.assertTrue(isAggregator.contains("YES") || isAggregator.contains("NO"));
+      // nothing to validate for signatures as it may be not be present if operandTypeChecker is not defined.
```
Comment (zachjsh) [inline sql/src/test/java/org/apache/druid/sql/calcite/schema/InformationSchemaTest.java:108]: Should we validate this field for one or a few known functions, that do have the operandTypeChecker defined, either in this test or another?

Comment (zachjsh) [reply sql/src/test/java/org/apache/druid/sql/calcite/schema/InformationSchemaTest.java:108]: ah nvm, I see you have that in the test below, nice!

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/InformationSchema.java (hunk 1 above)
Comment (abhishekrb19) [reply sql/src/main/java/org/apache/druid/sql/calcite/schema/InformationSchema.java]: Yeah, good point - I was basically using the inverse of `DruidOperatorTable#normalizeSyntax` to only consider function syntax. I've added a static function `DruidOperatorTable#isFunctionSyntax()` that gets used by `normalizeSyntax` and inside the routine table scan here. 

Also, it seems like we don't have test coverage for `DruidOperatorTable`, so I added a unit test class `DruidOperatorTableTest` to cover some basic functionality including the new function added. Other existing functionality in the class is already covered by Calcite tests indirectly.

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/InformationSchema.java (hunk 2 above)
Comment (abhishekrb19) [reply sql/src/main/java/org/apache/druid/sql/calcite/schema/InformationSchema.java]: thanks, added.

**Review by zachjsh — APPROVED** [review]
🚀

**Pull request merged**